channels/rdpgfx/server/rdpgfx_main.c

```diff
@@ -0,0 +1,1659 @@
+	return rdpgfx_server_single_packet_send(context, s);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_recv_frame_acknowledge_pdu(RdpgfxServerContext* context,
+		wStream* s)
+{
+	RDPGFX_FRAME_ACKNOWLEDGE_PDU pdu;
+	UINT error = CHANNEL_RC_OK;
+
+	if (Stream_GetRemainingLength(s) < sizeof(RDPGFX_FRAME_ACKNOWLEDGE_PDU))
+	{
+		WLog_ERR(TAG, "not enough data!");
+		return ERROR_INVALID_DATA;
+	}
+
+	Stream_Read_UINT32(s, pdu.queueDepth); /* queueDepth (4 bytes) */
+	Stream_Read_UINT32(s, pdu.frameId); /* frameId (4 bytes) */
+	/* totalFramesDecoded (4 bytes) */
+	Stream_Read_UINT32(s, pdu.totalFramesDecoded);
+
+	if (context)
+	{
+		IFCALLRET(context->FrameAcknowledge, error, context, &pdu);
+
+		if (error)
+			WLog_ERR(TAG, "context->FrameAcknowledge failed with error %lu",
+					 error);
+	}
+
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_recv_cache_import_offer_pdu(RdpgfxServerContext* context,
+		wStream* s)
+{
+	UINT16 index;
+	RDPGFX_CACHE_IMPORT_OFFER_PDU pdu;
+	RDPGFX_CACHE_ENTRY_METADATA* cacheEntries;
+	UINT error = CHANNEL_RC_OK;
+
+	if (Stream_GetRemainingLength(s) < sizeof(UINT16))
+	{
+		WLog_ERR(TAG, "not enough data!");
+		return ERROR_INVALID_DATA;
+	}
+
+	/* cacheEntriesCount (2 bytes) */
+	Stream_Read_UINT16(s, pdu.cacheEntriesCount);
+
+	if (pdu.cacheEntriesCount <= 0)
+	{
+		/* According to the latest spec, capsSetCount <= 3 */
+		WLog_ERR(TAG, "Invalid cacheEntriesCount: %u", pdu.cacheEntriesCount);
+		return ERROR_INVALID_DATA;
+	}
+
+	if (Stream_GetRemainingLength(s) <
+			(pdu.cacheEntriesCount * sizeof(RDPGFX_CACHE_ENTRY_METADATA)))
+	{
+		WLog_ERR(TAG, "not enough data!");
+		return ERROR_INVALID_DATA;
+	}
+
+	pdu.cacheEntries = (RDPGFX_CACHE_ENTRY_METADATA*)
+					   calloc(pdu.cacheEntriesCount,
+							  sizeof(RDPGFX_CACHE_ENTRY_METADATA));
+
+	if (!pdu.cacheEntries)
+	{
+		WLog_ERR(TAG, "calloc failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	for (index = 0; index < pdu.cacheEntriesCount; index++)
+	{
+		cacheEntries = &(pdu.cacheEntries[index]);
+		Stream_Read_UINT64(s, cacheEntries->cacheKey); /* cacheKey (8 bytes) */
+		/* bitmapLength (4 bytes) */
+		Stream_Read_UINT32(s, cacheEntries->bitmapLength);
+	}
+
+	if (context)
+	{
+		IFCALLRET(context->CacheImportOffer, error, context, &pdu);
+
+		if (error)
+			WLog_ERR(TAG, "context->CacheImportOffer failed with error %lu",
+					 error);
+	}
+
+	free(pdu.cacheEntries);
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpgfx_recv_caps_advertise_pdu(RdpgfxServerContext* context,
+		wStream* s)
+{
+	UINT16 index;
+	RDPGFX_CAPSET* capsSet;
+	RDPGFX_CAPSET capsSets[3];
+	RDPGFX_CAPS_ADVERTISE_PDU pdu;
+	UINT error = CHANNEL_RC_OK;
+	UINT32 capsDataLength;
+
+	if (Stream_GetRemainingLength(s) < sizeof(UINT16))
```

Comment:
Just don't want to introduce any magic number in code. It's OK to change it to 2 as 16bit exactly mean 2 bytes  ;)